src/github.com/getlantern/errlog/errlog.go

```diff
@@ -0,0 +1,540 @@
+	"os"
+	"os/exec"
+	"reflect"
+	"runtime"
+	"strconv"
+	"syscall"
+	"time"
+
+	"github.com/getlantern/golog"
+	"github.com/getlantern/jibber_jabber"
+	"github.com/getlantern/osversion"
+)
+
+type systemInfo struct {
+	OSType    string `json:"osType"`
+	OSVersion string `json:"osVersion"`
+	OSArch    string `json:"osArch"`
+}
+
+func (si *systemInfo) String() string {
+	var buf bytes.Buffer
+	if si.OSType != "" {
+		_, _ = buf.WriteString(" OSType=" + si.OSType)
+	}
+	if si.OSVersion != "" {
+		_, _ = buf.WriteString(" OSVersion=\"" + si.OSVersion + "\"")
+	}
+	if si.OSArch != "" {
+		_, _ = buf.WriteString(" OSArch=" + si.OSArch)
+	}
+	return buf.String()
+}
+
+type UserLocale struct {
+	TimeZone string `json:"timeZone,omitempty"`
+	Language string `json:"language,omitempty"`
+	Country  string `json:"country,omitempty"`
+}
+
+func (si *UserLocale) String() string {
+	var buf bytes.Buffer
+	if si.TimeZone != "" {
+		_, _ = buf.WriteString(" TimeZone=" + si.TimeZone)
+	}
+	if si.Language != "" {
+		_, _ = buf.WriteString(" Language=" + si.Language)
+	}
+	if si.Country != "" {
+		_, _ = buf.WriteString(" Country=" + si.Country)
+	}
+	return buf.String()
+}
+
+// ProxyType is the type of various proxy channel
+type ProxyType string
+
+const (
+	// direct access, no proxying at all
+	NoProxy ProxyType = "no"
+	// access through Lantern hosted chained server
+	ChainedProxy ProxyType = "chained"
+	// access through domain fronting
+	FrontedProxy ProxyType = "fronted"
+	// access through direct domain fronting
+	DirectFrontedProxy ProxyType = "DDF"
+)
+
+// ProxyingInfo encapsulates fields to describe an access through a proxy channel.
+type ProxyingInfo struct {
+	ProxyType  ProxyType `json:"proxyType,omitempty"`
+	LocalAddr  string    `json:"localAddr,omitempty"`
+	ProxyAddr  string    `json:"proxyAddr,omitempty"`
+	ProxyDC    string    `json:"proxyDataCenter,omitempty"`
+	OriginSite string    `json:"originSite,omitempty"`
+	Scheme     string    `json:"scheme,omitempty"`
+}
+
+func (pi *ProxyingInfo) String() string {
+	var buf bytes.Buffer
+	if pi.ProxyType != "" {
+		_, _ = buf.WriteString(" ProxyType=" + string(pi.ProxyType))
+	}
+	if pi.LocalAddr != "" {
+		_, _ = buf.WriteString(" LocalAddr=" + pi.LocalAddr)
+	}
+	if pi.ProxyAddr != "" {
+		_, _ = buf.WriteString(" ProxyAddr=" + pi.ProxyAddr)
+	}
+	if pi.ProxyDC != "" {
+		_, _ = buf.WriteString(" ProxyDC=" + pi.ProxyDC)
+	}
+	if pi.OriginSite != "" {
+		_, _ = buf.WriteString(" OriginSite=" + pi.OriginSite)
+	}
+	if pi.Scheme != "" {
+		_, _ = buf.WriteString(" Scheme=" + pi.Scheme)
+	}
+	return buf.String()
+}
+
+// UserAgentInfo encapsulates traits of the browsers or 3rd party applications
+// directing traffic through Lantern.
+type UserAgentInfo struct {
+	UserAgent string `json:"userAgent,omitempty"`
+}
+
+func (ul *UserAgentInfo) String() string {
+	return fmt.Sprintf("UserAgent=%s", ul.UserAgent)
+}
+
+// Error wraps system and application errors in unified structure
+type Error struct {
```

Comment:
The problem is that we can't predefine all possible codes. It's also tedious for the caller to find the correct error codes. For monitoring purpose, most operations are not to exactly match a few error codes, but list the top N errors, show the distribution, or send alert if the count exceeds some criteria. For troubleshooting, we have user id, device id, request id, ip address, etc to acurately filter the dataset.

Changed field value does little harm as they are same in a version.